### PR TITLE
[fbpick-coarse] Replace train/validation datasets with global-anchor datasets

### LINKED ISSUE #27: [fbpick-coarse] Replace train/validation datasets with global-anchor datasets
Background

fbpick-coarse を global-anchor resize mode に移行するため、学習・validation 用 dataset を従来の local crop 方式から、full gather 全体を使う固定サイズ入力方式へ差し替える。

従来の coarse dataset は、おおむね以下の流れだった。

SEG-Y gather
→ trace window crop
→ first-break が入るように time crop / pad
→ model input

新しい coarse-global 設計では、以下の流れに変更する。

SEG-Y full gather
→ gap-aware trace anchor selection
→ selected anchor traces
→ time resampling to Wc=2048
→ fixed input shape (3, 256, 2048)
→ model input

coarse stage の目的は、sample 単位の精密な picking ではなく、後段の fine / physics stage が参照する大まかな first-break 位置を推定することにある。
そのため、train / validation dataset でも tiled / crop 前提ではなく、gather 全体の大局的構造を保持する global-anchor dataset を使う。

Goal

fbpick-coarse の train / validation dataset を、global-anchor resize mode に対応した dataset へ置き換える。

固定契約:

input shape:  (3, 256, 2048)
target shape: (1, 256, 2048)

channels:
  0: waveform
  1: offset_ch
  2: time_ch

train と validation の違い:

train:
  gap-aware random anchor selection

validation:
  gap-aware center anchor selection
Prerequisites

この issue は、以下の issue が完了済みであることを前提にする。

[fbpick-coarse] Add global-anchor 256×2048 config and checkpoint contract
[dataset] Add full-gather sampling support
[fbpick-coarse] Implement gap-aware trace anchor selection
[fbpick-coarse] Implement time resampling and FB coordinate projection

前提として利用できる API:

TraceSubsetSampler.draw_full_gather
select_gap_aware_trace_anchors
resample_waveform_time_axis
build_coarse_time_grid
project_fb_indices_to_coarse_time
build_coarse_fb_labels_for_anchors

実際の関数名は実装済みコードに合わせる。

Scope
1. Add global-anchor train dataset for fbpick-coarse

対象候補:

packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/build_dataset.py

新しい train dataset を追加する。

想定名:

GlobalAnchorCoarseTrainDataset

または既存命名規則に合わせる。

dataset の流れ:

1. SEG-Y file info を選ぶ
2. TraceSubsetSampler.draw_full_gather で primary gather 全体の trace indices を取得する
3. full gather の offset / first-break label を取得する
4. gap-aware random anchor selection を行う
5. anchor trace waveform を読み込む
6. waveform を time axis 方向に Wc=2048 へ resample する
7. waveform を per-trace standardize する
8. offset_ch と time_ch を作る
9. raw first-break index を coarse grid 上の index へ投影する
10. target map を作る
11. input / target / meta を返す

返す tensor shape:

input:  (3, 256, 2048)
target: (1, 256, 2048)
2. Add deterministic global-anchor validation dataset

validation dataset は train dataset と同じ input construction を使う。
ただし anchor selection は random ではなく center mode にする。

想定名:

GlobalAnchorCoarseValidationDataset

または train dataset に anchor_mode を渡して共通化する。

期待する違い:

train:
  anchor_mode = "random"
  rng を使う

validation:
  anchor_mode = "center"
  rng に依存しない

validation dataset は、同じ sample を何回読んでも同じ anchor mapping を返す必要がある。

3. Replace build_train_dataset

対象候補:

packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/build_dataset.py

build_train_dataset を global-anchor dataset を返すように変更する。

旧設計の local crop dataset ではなく、以下の dataset を返す。

GlobalAnchorCoarseTrainDataset

config は以下を参照する。

transform.trace_len == 256
transform.time_len == 2048
trace_anchor.train_mode == "random"
trace_anchor.gap_ratio
trace_anchor.min_gap_m

transform.trace_len / transform.time_len が固定契約と一致しない場合は即エラーにする。
この validation が既に config 層で行われている場合は、重複せず既存 validation を使う。

4. Replace build_labeled_infer_dataset or validation builder

対象候補:

packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/build_dataset.py

既存の labeled validation / labeled infer dataset builder を global-anchor validation dataset に差し替える。

返す dataset は以下の仕様にする。

input shape:  (3, 256, 2048)
target shape: (1, 256, 2048)
anchor mode:  center

validation の purpose は、training 中の coarse model 評価である。
したがって、local crop ではなく、inference と同じ deterministic global-anchor 入力で評価する。

5. Preserve input channel contract

input は 3 channel を維持する。

waveform
offset_ch
time_ch
waveform
shape: (256, 2048)
dtype: float32

anchor trace waveform を time resample したあと、既存仕様に合わせて per-trace standardize する。

pad row はゼロ埋めにする。

offset_ch
shape: (256, 2048)
dtype: float32

各 anchor row の offset を time 方向に broadcast する。

pad row はゼロ埋めにする。
ただし pad row は trace_valid=False で識別できるように meta に残す。

既存 coarse input が offset を正規化している場合は、同じ正規化規則を使う。
この issue で offset のスケールを意味的に変更しない。

time_ch
shape: (256, 2048)
dtype: float32

build_coarse_time_grid で作った coarse time grid を trace 方向に broadcast する。

既存 coarse input が秒単位 time channel を使っている場合は秒単位を維持する。
既存 coarse input が正規化 time channel を使っている場合は同じ規則に合わせる。

6. Build target map on coarse grid

target は coarse grid 上に作る。

target shape: (1, 256, 2048)

raw first-break sample index は、以下の変換で coarse sample index へ投影する。

raw sample index → coarse sample index

座標変換は、既存の time projection utility を必ず使う。
式を dataset 内に直接重複実装しない。

pad row または invalid first-break row は、target map 生成時に無効行として扱う。

target map の作り方は既存 coarse dataset の仕様に合わせる。

例:

Gaussian target
one-hot target
ignore mask

既存 build_plan.py や target builder が使える場合は、重複実装せずそれを再利用する。

7. Add dataset metadata for downstream debugging

dataset item には、可能であれば以下の metadata を含める。

trace_valid
anchor_raw_indices
anchor_source_pos
anchor_offsets_m
segment_id
segments
raw_time_len
coarse_time_len
dt_sec
dt_eff_sec
raw_to_coarse_factor
coarse_to_raw_factor
fb_idx_raw_for_anchors
fb_idx_coarse_for_anchors

metadata は tensor 化が必要なものと debug 用 object を分けてよい。

DataLoader の collate で扱いづらい object がある場合は、以下のどちらかにする。

- numpy array / torch tensor に正規化する
- debug metadata は最小限にする

最低限、tests で anchor selection と coordinate projection を確認できる metadata は保持する。

8. Ensure pad rows are handled correctly

N < 256 の gather では pad row が発生する。

期待する挙動:

waveform pad rows are zero
offset_ch pad rows are zero
target pad rows are ignored or zero according to existing target convention
trace_valid is False for pad rows
fb_idx_coarse_for_anchors is ignore_index for pad rows

pad row が loss に悪影響を与えないようにする。

既存 loss が ignore mask を扱えない場合は、この issue 内で target convention を確認し、pad row が positive label を持たないようにする。
必要であれば pad row loss mask 対応は別 issue として切り出す。

9. Keep old local-crop dataset code only if needed

旧 local-crop dataset をすぐ削除する必要はない。
ただし、fbpick-coarse の default path は global-anchor dataset に切り替える。

旧 dataset を残す場合でも、以下を満たすこと。

- 新 config の global_anchor_resize path では旧 dataset が使われない
- tests で新 path が明示的に検証される
- 旧 path を残す理由がコード上または docs で分かる

後方互換のために複雑な分岐を増やさない。
不要になった旧 path の削除は別 issue で扱ってよい。

Out of scope

この issue では以下は実装しない。

- raw inference dataset の追加
- tiled-W inference の削除
- model forward / infer loop の変更
- anchor prediction の元 trace 座標への復元
- .coarse.npz 出力変更
- checkpoint metadata validation の追加
- trace anchor selection utility 自体の実装
- time resampling utility 自体の実装

この issue は、既に実装済みの utility を組み合わせて、train / validation dataset を global-anchor 化することに集中する。

Acceptance criteria
fbpick-coarse の train dataset が global-anchor input を返す。
train input shape が (3, 256, 2048) である。
train target shape が (1, 256, 2048) である。
train dataset は draw_full_gather を使い、local trace window crop を行わない。
train dataset は mode="random" の gap-aware anchor selection を使う。
validation dataset が global-anchor input を返す。
validation input shape が (3, 256, 2048) である。
validation target shape が (1, 256, 2048) である。
validation dataset は mode="center" の gap-aware anchor selection を使う。
validation dataset は同じ item に対して deterministic である。
waveform / offset_ch / time_ch の 3 channel contract が維持されている。
raw first-break label が coarse grid 上へ正しく投影される。
pad row が trace_valid=False として扱われる。
pad row が target positive label を持たない。
既存の coarse train loop が dataset を読める。
旧 local-crop coarse dataset が default path で使われない。
unit tests / smoke tests が追加または更新されている。
既存 tests が通る。
Suggested tests
1. Train dataset returns fixed global-anchor shape

小さな test SEG-Y fixture または mock file info を使う。

期待:

sample["input"].shape == (3, 256, 2048)
sample["target"].shape == (1, 256, 2048)
2. Validation dataset returns fixed global-anchor shape

期待:

sample["input"].shape == (3, 256, 2048)
sample["target"].shape == (1, 256, 2048)
3. Train dataset uses random anchors

同じ dataset item を、異なる rng seed または異なる epoch context で取得する。

期待:

anchor_source_pos が一部変わり得る
input shape は変わらない
target shape は変わらない

同じ seed で再現性が必要な設計なら、同じ seed で一致することも確認する。

4. Validation dataset uses deterministic anchors

同じ validation item を複数回取得する。

期待:

anchor_source_pos が完全一致する
input が完全一致または数値誤差範囲内で一致する
target が完全一致または数値誤差範囲内で一致する
5. Dataset uses full gather path

TraceSubsetSampler.draw_full_gather を呼んでいることを確認する。

方法例:

- mock sampler で draw_full_gather の呼び出しを検証する
- local crop では返り得ない trace 数や anchor metadata を検証する

期待:

subset window crop path が使われていない
6. Pad rows are validly masked

N < 256 の gather を使う。

期待:

trace_valid[:N].all()
not trace_valid[N:].any()
waveform pad rows are zero
target pad rows have no positive label
fb_idx_coarse_for_anchors[N:] == ignore_index
7. FB coordinate projection is used

既知の raw first-break index を持つ fixture を使う。

例:

raw_time_len = 6016
coarse_time_len = 2048
fb_idx_raw = [0, 6015]

期待:

fb_idx_coarse = [0, 2047]

dataset 内の metadata または target map peak から確認する。

8. Offset channel matches selected anchors

anchor offset が既知の fixture を使う。

期待:

offset_ch[row, :] が anchor_offsets_m[row] に対応している
pad row の offset_ch は 0

既存仕様で offset 正規化が入る場合は、その正規化後の値で確認する。

9. Time channel endpoint alignment

期待:

time_ch[:, 0] == 0.0
time_ch[:, -1] == (raw_time_len - 1) * dt_sec

既存仕様で time 正規化が入る場合は、既存仕様に合わせた期待値にする。

10. Coarse train smoke test

最小 config で fbpick-coarse train dataset が DataLoader 経由で batch 化できることを確認する。

期待:

batch input shape == (B, 3, 256, 2048)
batch target shape == (B, 1, 256, 2048)
Implementation notes

推奨する実装構成:

build_dataset.py

- GlobalAnchorCoarseDataset
  - split train / validation by anchor_mode
  - train uses anchor_mode="random"
  - validation uses anchor_mode="center"

- build_train_dataset
  - returns GlobalAnchorCoarseDataset(anchor_mode="random")

- build_labeled_infer_dataset
  - returns GlobalAnchorCoarseDataset(anchor_mode="center")

dataset 内で直接実装しない方がよい処理:

- offset gap 判定
- anchor selection
- raw/coarse time coordinate conversion
- waveform time resampling

これらは既存 utility を呼び出す。

座標変換式を dataset 内に重複して書かない。
特に raw sample index と coarse sample index の相互変換は、後続 inference / .coarse.npz 保存でも使うため、共通 utility に一本化する。

### PR DESCRIPTION
**Pull request by ozw4**
Closes #27

## Summary
- [fbpick-coarse] Replace train/validation datasets with global-anchor datasets

## Changed files
- `cli/_entrypoint.py`
- `packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/__init__.py`
- `packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/build_dataset.py`
- `packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/train.py`
- `packages/seisai-engine/tests/test_fbpick_coarse.py`

## Checks
- `.work/codex/checks.log`: 747 passed, 5 warnings in 56.83s

## Review
- `.work/codex/review.txt`: accept: yes
- findings: blocker 0, major 0, minor 1

### CODE REVIEW

**Comment by ozw4** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/ozw4/seisai/pull/29#pullrequestreview-4185309302)
